mlc-ai/mlc-llm
# [SLM] Add support for Chatglm3 architecture

This pr enable Chatglm3 model.

QA Examples:
```
tlopex@tlopex-OMEN-by-HP-Laptop-17-ck1xxx:~/mlc-llm$ mlc_llm chat dist/chatglm3-6b-q4f16_1-MLC              --device "cuda:0" --overrides context_window_size=8192              --model-lib-path ./dist/libs/chatglm3-6b-q4f16_1-cuda.so
[2024-04-06 15:01:30] INFO config.py:106: Overriding context_window_size from None to 8192
[2024-04-06 15:01:31] INFO auto_device.py:76: Found device: cuda:0
[2024-04-06 15:01:31] INFO chat_module.py:379: Using model folder: /home/tlopex/mlc-llm/dist/chatglm3-6b-q4f16_1-MLC
[2024-04-06 15:01:31] INFO chat_module.py:380: Using mlc chat config: /home/tlopex/mlc-llm/dist/chatglm3-6b-q4f16_1-MLC/mlc-chat-config.json
[2024-04-06 15:01:31] INFO chat_module.py:529: Using library model: ./dist/libs/chatglm3-6b-q4f16_1-cuda.so
[2024-04-06 15:01:32] INFO model_metadata.py:96: Total memory usage: 6515.82 MB (Parameters: 3349.82 MB. KVCache: 0.00 MB. Temporary buffer: 3166.00 MB)
[2024-04-06 15:01:32] INFO model_metadata.py:105: To reduce memory usage, tweak `prefill_chunk_size`, `context_window_size` and `sliding_window_size`
[15:01:32] /home/tlopex/mlc-llm/cpp/tokenizers.cc:79: Warning: Using `tokenizer.model` since we cannot locate `tokenizer.json`.
It is recommended to use `tokenizer.json` to ensure all token mappings are included, since currently, files like `added_tokens.json`, `tokenizer_config.json` are ignored.
Consider converting `tokenizer.model` to `tokenizer.json` by compiling the model with MLC again, or see if MLC's huggingface provides this file.
You can use the following special commands:
  /help               print the special commands
  /exit               quit the cli
  /stats              print out the latest stats (token/sec)
  /reset              restart a fresh chat
  /set [overrides]    override settings in the generation config. For example,
                      `/set temperature=0.5;max_gen_len=100;stop=end,stop`
                      Note: Separate stop words in the `stop` option with commas (,).
  Multi-line input: Use escape+enter to start a new line.

问: 你好
答: 
您好,请问有什么需要帮助的吗?
问: Can you speak English?
答: 
Yes, I can speak English.
```


MLC server test:
```
# test_server_rest_api.py
import requests
import json

model = "dist/chatglm3-6b-q4f16_1-MLC"
port = 8000
payload = {
    "model": f"{model}",
    "messages": [{"role": "user", "content": "Hello!"}],
    "stream": False,
}

r = requests.post(f"http://127.0.0.1:{port}/v1/chat/completions", json=payload)
if r.status_code != 200:
    print(r.json())
else:
    print(r.json()["choices"][0]["message"]["content"])
```
States:
```
tlopex@tlopex-OMEN-by-HP-Laptop-17-ck1xxx:~/mlc-llm$ python -m mlc_llm.serve.server --model dist/chatglm3-6b-q4f16_1-MLC --model-lib-path ./dist/libs/chatglm3-6b-q4f16_1-cuda.so
[2024-04-06 14:36:06] INFO auto_device.py:76: Found device: cuda:0
[2024-04-06 14:36:07] INFO auto_device.py:85: Not found device: rocm:0
[2024-04-06 14:36:08] INFO auto_device.py:85: Not found device: metal:0
[2024-04-06 14:36:09] INFO auto_device.py:85: Not found device: vulkan:0
[2024-04-06 14:36:09] INFO auto_device.py:85: Not found device: opencl:0
[2024-04-06 14:36:09] INFO auto_device.py:33: Using device: cuda:0
[2024-04-06 14:36:09] INFO chat_module.py:379: Using model folder: /home/tlopex/mlc-llm/dist/chatglm3-6b-q4f16_1-MLC
[2024-04-06 14:36:09] INFO chat_module.py:380: Using mlc chat config: /home/tlopex/mlc-llm/dist/chatglm3-6b-q4f16_1-MLC/mlc-chat-config.json
[2024-04-06 14:36:09] INFO chat_module.py:529: Using library model: ./dist/libs/chatglm3-6b-q4f16_1-cuda.so
[2024-04-06 14:36:11] INFO engine.py:246: Estimated KVCacheConfig "max_total_sequence_length": 277728.
[2024-04-06 14:36:11] INFO engine.py:251: Estimated total single GPU memory usage: 14504.99 MB (Parameters: 3349.82 MB. KVCache: 7780.53 MB. Temporary buffer: 3374.65 MB)
[14:36:11] /home/tlopex/mlc-llm/cpp/tokenizers.cc:79: Warning: Using `tokenizer.model` since we cannot locate `tokenizer.json`.
It is recommended to use `tokenizer.json` to ensure all token mappings are included, since currently, files like `added_tokens.json`, `tokenizer_config.json` are ignored.
Consider converting `tokenizer.model` to `tokenizer.json` by compiling the model with MLC again, or see if MLC's huggingface provides this file.
[14:36:11] /home/tlopex/mlc-llm/cpp/tokenizers.cc:79: Warning: Using `tokenizer.model` since we cannot locate `tokenizer.json`.
It is recommended to use `tokenizer.json` to ensure all token mappings are included, since currently, files like `added_tokens.json`, `tokenizer_config.json` are ignored.
Consider converting `tokenizer.model` to `tokenizer.json` by compiling the model with MLC again, or see if MLC's huggingface provides this file.
INFO:     Started server process [16670]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     127.0.0.1:38704 - "POST /v1/chat/completions HTTP/1.1" 200 OK
INFO:     127.0.0.1:51360 - "POST /v1/chat/completions HTTP/1.1" 200 OK
```
